File: src/world/World.js
/**
 * World Class
 * Manages the game world grid and tiles
 */
class World {
    constructor(width, height, tileWidth = 64, tileHeight = 32, assetLoader = null) {
        this.width = width;
        this.height = height;
        this.tileWidth = tileWidth;
        this.tileHeight = tileHeight;
        this.assetLoader = assetLoader;
        this.tiles = [];
        this.entities = [];
        
        this.generate();
    }
    
    /**
     * Generate the world
     */
    generate() {
        // Generate biome map
        this.biomeMap = this.generateBiomeMap();
        
        for (let y = 0; y < this.height; y++) {
            this.tiles[y] = [];
            for (let x = 0; x < this.width; x++) {
                // Create varied terrain based on biome
                const biome = this.biomeMap[y][x];
                const type = this.generateTileType(x, y, biome);
                this.tiles[y][x] = new Tile(x, y, type);
            }
        }
        
        // Add decorations after all tiles are created
        this.generateDecorations();
    }
    
    /**
     * Generate biome map using simple noise-like algorithm
     */
    generateBiomeMap() {
        const biomeMap = [];
        const biomeTypes = [
            Biome.TYPES.FOREST,
            Biome.TYPES.PLAINS,
            Biome.TYPES.DESERT,
            Biome.TYPES.MOUNTAINS,
            Biome.TYPES.WETLANDS
        ];
        
        for (let y = 0; y < this.height; y++) {
            biomeMap[y] = [];
            for (let x = 0; x < this.width; x++) {
                // Use simple region-based biome assignment
                const regionX = Math.floor(x / 10);
                const regionY = Math.floor(y / 10);
                const biomeIndex = (regionX + regionY * 3) % biomeTypes.length;
                biomeMap[y][x] = new Biome(biomeTypes[biomeIndex]);
            }
        }
        
        return biomeMap;
    }
    
    /**
     * Generate tile type based on position and biome
     */
    generateTileType(x, y, biome) {
        // Check for water first
        if (biome.shouldSpawnWater()) {
            return Tile.TYPES.WATER;
        }
        
        // Use biome's primary or secondary tile
        const random = Math.random();
        if (random < 0.8) {
            return biome.getPrimaryTile();
        } else {
            return biome.getSecondaryTile();
        }
    }
    
    /**
     * Generate decorations for tiles
     */
    generateDecorations() {
        // Decoration type counts - now using more variations from individual assets
        const TREE_TYPES = 20; // Using 20 tree variations
        const BUSH_TYPES = 3;
        const ROCK_TYPES = 2;
        
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                const tile = this.tiles[y][x];
                const biome = this.biomeMap[y][x];
                
                // Skip water tiles
                if (tile.type === Tile.TYPES.WATER) {
                    // Add pond decorations to some water tiles
                    if (Math.random() < 0.3) {
                        tile.setDecoration('pond');
                    }
                    continue;
                }
                
                // Add trees based on biome
                if (biome.shouldSpawnTree() && tile.type.walkable) {
                    const treeType = Math.floor(Math.random() * TREE_TYPES);
                    tile.setDecoration(`tree_${treeType}`);
                    tile.isResource = true; // Mark as gatherable resource
                }
                // Add bushes based on biome
                else if (biome.shouldSpawnBush() && tile.type.walkable) {
                    const bushType = Math.floor(Math.random() * BUSH_TYPES) + 1;
                    tile.setDecoration(`bush_${bushType}`);
                }
                // Add rocks based on biome
                else if (biome.shouldSpawnRock() && tile.type.walkable) {
                    const rockType = Math.floor(Math.random() * ROCK_TYPES) + 1;
                    tile.setDecoration(`rocks_${rockType}`);
                    tile.isResource = true; // Mark as gatherable resource
                }
            }
        }
    }
    
    /**
     * Get tile at position
     */
    getTile(x, y) {
        if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
            return this.tiles[y][x];
        }
        return null;
    }
    
    /**
     * Get tile at screen position
     */
    getTileAtScreen(screenX, screenY, camera) {
        const adjustedX = screenX + camera.x;
        const adjustedY = screenY + camera.y;
        const tileCoords = IsometricUtils.screenToTile(
            adjustedX,
            adjustedY,
            this.tileWidth,
            this.tileHeight
        );
        return this.getTile(tileCoords.x, tileCoords.y);
    }
    
    /**
     * Add entity to world
     */
    addEntity(entity) {
        this.entities.push(entity);
    }
    
    /**
     * Remove entity from world
     */
    removeEntity(entity) {
        const index = this.entities.indexOf(entity);
        if (index > -1) {
            this.entities.splice(index, 1);
        }
    }
    
    /**
     * Update world
     */
    update(deltaTime) {
        // Update all entities
        for (const entity of this.entities) {
            if (entity.update) {
                entity.update(deltaTime, this);
            }
        }
    }
    
    /**
     * Render world
     */
    render(renderer, camera, isometricRenderer) {
        // Calculate visible tile range
        const startX = Math.max(0, Math.floor(-camera.x / this.tileWidth) - 5);
        const startY = Math.max(0, Math.floor(-camera.y / this.tileHeight) - 5);
        const endX = Math.min(this.width, startX + 40);
        const endY = Math.min(this.height, startY + 40);
        
        // Render tiles
        for (let y = startY; y < endY; y++) {
            for (let x = startX; x < endX; x++) {
                const tile = this.tiles[y][x];
                const screenPos = IsometricUtils.tileToScreen(
                    tile.x,
                    tile.y,
                    this.tileWidth,
                    this.tileHeight
                );
                
                // Try to use tile images if available
                let tileImage = null;
                if (this.assetLoader) {
                    // Map tile types to asset names with variations
                    let baseName = null;
                    if (tile.type === Tile.TYPES.GRASS) {
                        baseName = 'grass_green';
                    } else if (tile.type === Tile.TYPES.SAND) {
                        baseName = 'sand';
                    } else if (tile.type === Tile.TYPES.DIRT) {
                        baseName = 'dirt';
                    } else if (tile.type === Tile.TYPES.STONE) {
                        baseName = 'stone_path';
                    }
                    
                    // Use the tile's variation index for visual diversity
                    if (baseName) {
                        tileImage = this.assetLoader.getImage(`${baseName}_${tile.tileVariation}`);
                    }
                }
                
                // Draw tile with image or fallback to solid color
                if (tileImage) {
                    renderer.drawImage(
                        tileImage,
                        screenPos.x - this.tileWidth / 2 - camera.x,
                        screenPos.y - this.tileHeight / 2 - camera.y,
                        this.tileWidth,
                        this.tileHeight
                    );
                } else {
                    // Fallback to solid color tiles
                    isometricRenderer.drawIsometricTile(
                        screenPos.x,
                        screenPos.y,
                        this.tileWidth,
                        this.tileHeight,
                        tile.getColor(),
                        camera
                    );
                }
                
                // Draw decoration if present
                if (tile.decoration && this.assetLoader) {
                    const decorationImage = this.assetLoader.getImage(tile.decoration);
                    if (decorationImage) {
                        // Draw decoration centered on tile
                        const decorX = screenPos.x - decorationImage.width / 2 - camera.x;
                        const decorY = screenPos.y - decorationImage.height + this.tileHeight / 2 - camera.y;
                        renderer.drawImage(
                            decorationImage,
                            decorX,
                            decorY,
                            decorationImage.width,
                            decorationImage.height
                        );
                    }
                }
                
                // Draw building if present
                if (tile.building) {
                    tile.building.render(renderer, camera, isometricRenderer, screenPos);
                }
            }
        }
        
        // Render entities
        for (const entity of this.entities) {
            if (entity.render) {
                entity.render(renderer, camera, isometricRenderer);
            }
        }
    }
}

File: cpp/src/world/World.cpp
#include "world/World.h"
#include "world/Biome.h"
#include "rendering/Renderer.h"
#include "rendering/IsometricRenderer.h"
#include "rendering/Camera.h"
#include "utils/NoiseGenerator.h"
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <sstream>

World::World(int width, int height)
    : width(width)
    , height(height)
    , noiseGen(std::make_unique<NoiseGenerator>(static_cast<uint32_t>(std::time(nullptr))))
{
    // Initialize tiles
    tiles.resize(height);
    for (int y = 0; y < height; ++y) {
        tiles[y].resize(width);
        for (int x = 0; x < width; ++x) {
            tiles[y][x] = std::make_unique<Tile>(x, y, TileType::GRASS);
        }
    }
}

World::~World() {
}

void World::generate() {
    // Seed random number generator
    std::srand(static_cast<unsigned>(std::time(nullptr)));
    
    // Generate biome map first
    generateBiomeMap();
    
    // Generate terrain based on biomes
    generateTerrain();
    
    // Generate decorations
    generateDecorations();
    
    std::cout << "World generated: " << width << "x" << height << " tiles with biomes" << std::endl;
}

void World::update(float deltaTime) {
    (void)deltaTime; // Unused - reserved for future weather/day-night cycle
    // World update logic (future: weather, day/night cycle, etc.)
}

void World::render(Renderer* renderer, IsometricRenderer* isoRenderer, Camera* camera) {
    (void)renderer; // Unused - using isoRenderer for rendering
    (void)camera; // Unused - camera handled by renderer
    // Render tiles in isometric order (back to front, left to right)
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            const Tile* tile = getTile(x, y);
            if (tile) {
                isoRenderer->drawIsometricColoredTile(
                    x, y,
                    tile->getColor()
                );
            }
        }
    }
}

Tile* World::getTile(int x, int y) {
    if (!isValidPosition(x, y)) {
        return nullptr;
    }
    return tiles[y][x].get();
}

const Tile* World::getTile(int x, int y) const {
    if (!isValidPosition(x, y)) {
        return nullptr;
    }
    return tiles[y][x].get();
}

bool World::isValidPosition(int x, int y) const {
    return x >= 0 && x < width && y >= 0 && y < height;
}

bool World::loadFromFile(const char* filename) {
    // TODO: Implement JSON loading
    std::cout << "Loading world from: " << filename << std::endl;
    return false;
}

bool World::saveToFile(const char* filename) const {
    // TODO: Implement JSON saving
    std::cout << "Saving world to: " << filename << std::endl;
    return false;
}

void World::generateBiomeMap() {
    // Create biome map using noise-based temperature and moisture
    biomeMap.resize(height);
    
    const float scale = 0.05f; // Scale for noise (larger = bigger biomes)
    
    for (int y = 0; y < height; ++y) {
        biomeMap[y].resize(width);
        for (int x = 0; x < width; ++x) {
            // Generate temperature and moisture using different noise octaves
            float temperature = noiseGen->fractalNoise2D(x * scale, y * scale, 4, 0.5f);
            float moisture = noiseGen->fractalNoise2D(x * scale + 1000.0f, y * scale + 1000.0f, 4, 0.5f);
            
            // Determine biome based on temperature and moisture
            BiomeType biomeType = getBiomeFromNoise(temperature, moisture);
            biomeMap[y][x] = std::make_unique<Biome>(biomeType);
        }
    }
}

BiomeType World::getBiomeFromNoise(float temperature, float moisture) const {
    // Map temperature and moisture to biomes
    // Temperature: 0.0 (cold) -> 1.0 (hot)
    // Moisture: 0.0 (dry) -> 1.0 (wet)
    
    if (temperature < 0.3f) {
        // Cold regions
        if (moisture > 0.5f) {
            return BiomeType::WETLANDS; // Cold and wet
        } else {
            return BiomeType::MOUNTAINS; // Cold and dry
        }
    } else if (temperature < 0.6f) {
        // Temperate regions
        if (moisture > 0.6f) {
            return BiomeType::FOREST; // Temperate and wet
        } else if (moisture > 0.3f) {
            return BiomeType::PLAINS; // Temperate and moderate
        } else {
            return BiomeType::DESERT; // Temperate and dry
        }
    } else {
        // Hot regions
        if (moisture > 0.5f) {
            return BiomeType::FOREST; // Hot and wet (tropical)
        } else {
            return BiomeType::DESERT; // Hot and dry
        }
    }
}

void World::generateTerrain() {
    // Generate terrain based on biomes and additional noise
    const float detailScale = 0.15f; // Finer detail for terrain variation
    
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            const Biome* biome = biomeMap[y][x].get();
            
            // Add detail noise for within-biome variation
            float detailNoise = noiseGen->noise2D(x * detailScale, y * detailScale);
            
            TileType tileType;
            
            // Check for water using noise (creates lakes and rivers)
            if (biome->shouldSpawnWater()) {
                // Use noise to create connected water bodies
                float waterNoise = noiseGen->fractalNoise2D(x * 0.08f, y * 0.08f, 3, 0.6f);
                if (waterNoise < 0.35f) {
                    tileType = TileType::WATER;
                } else {
                    // Use biome's primary or secondary tile based on detail noise
                    tileType = (detailNoise < 0.7f) ? biome->getPrimaryTile() : biome->getSecondaryTile();
                }
            } else {
                // Use biome's primary or secondary tile based on detail noise
                tileType = (detailNoise < 0.8f) ? biome->getPrimaryTile() : biome->getSecondaryTile();
            }
            
            tiles[y][x]->setType(tileType);
        }
    }
}

void World::generateDecorations() {
    const int TREE_TYPES = 20;
    const int BUSH_TYPES = 3;
    const int ROCK_TYPES = 2;
    
    // Use different noise frequencies for different decoration types
    const float treeScale = 0.2f;
    const float bushScale = 0.25f;
    const float rockScale = 0.18f;
    
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            Tile* tile = tiles[y][x].get();
            const Biome* biome = biomeMap[y][x].get();
            
            // Skip water tiles (add pond decorations to some)
            if (tile->getType() == TileType::WATER) {
                float pondNoise = noiseGen->noise2D(x * 0.3f, y * 0.3f);
                if (pondNoise > 0.7f) {
                    tile->setDecoration("pond");
                }
                continue;
            }
            
            // Skip non-walkable tiles
            if (!tile->isWalkable()) {
                continue;
            }
            
            // Use noise to create clustered decorations (more realistic)
            float treeNoise = noiseGen->fractalNoise2D(x * treeScale + 500.0f, y * treeScale + 500.0f, 2, 0.4f);
            float bushNoise = noiseGen->fractalNoise2D(x * bushScale + 1500.0f, y * bushScale + 1500.0f, 2, 0.4f);
            float rockNoise = noiseGen->fractalNoise2D(x * rockScale + 2500.0f, y * rockScale + 2500.0f, 2, 0.4f);
            
            // Combine biome probability with noise for natural clustering
            bool shouldPlaceTree = biome->shouldSpawnTree() && (treeNoise > 0.55f);
            bool shouldPlaceBush = biome->shouldSpawnBush() && (bushNoise > 0.6f);
            bool shouldPlaceRock = biome->shouldSpawnRock() && (rockNoise > 0.58f);
            
            // Place decorations (priority: trees > rocks > bushes)
            if (shouldPlaceTree) {
                int treeType = static_cast<int>(treeNoise * TREE_TYPES) % TREE_TYPES;
                std::stringstream ss;
                ss << "tree_" << treeType;
                tile->setDecoration(ss.str());
                tile->setResource(true);
            } else if (shouldPlaceRock) {
                int rockType = (static_cast<int>(rockNoise * ROCK_TYPES) % ROCK_TYPES) + 1;
                std::stringstream ss;
                ss << "rocks_" << rockType;
                tile->setDecoration(ss.str());
                tile->setResource(true);
            } else if (shouldPlaceBush) {
                int bushType = (static_cast<int>(bushNoise * BUSH_TYPES) % BUSH_TYPES) + 1;
                std::stringstream ss;
                ss << "bush_" << bushType;
                tile->setDecoration(ss.str());
            }
        }
    }
}
